posthog/hogql/database/schema/channel_type.py

```diff
@@ -38,29 +56,130 @@ def create_initial_domain_type(name: str):
     )
 
 
-def create_initial_channel_type(name: str):
+def create_initial_channel_type(name: str, custom_rules: Optional[list[CustomChannelRule]] = None):
     return ExpressionField(
         name=name,
         expr=create_channel_type_expr(
-            campaign=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_campaign"])]),
-            medium=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_medium"])]),
-            source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_source"])]),
-            referring_domain=ast.Call(
-                name="toString", args=[ast.Field(chain=["properties", "$initial_referring_domain"])]
+            source_exprs=ChannelTypeExprs(
+                campaign=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_campaign"])]),
+                medium=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_medium"])]),
+                source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_utm_source"])]),
+                referring_domain=ast.Call(
+                    name="toString", args=[ast.Field(chain=["properties", "$initial_referring_domain"])]
+                ),
+                gclid=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gclid"])]),
+                gad_source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gad_source"])]),
             ),
-            gclid=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gclid"])]),
-            gad_source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gad_source"])]),
+            custom_rules=custom_rules,
         ),
     )
 
 
+def custom_condition_to_expr(
```

Comment:
This is very similar to the code in hogql/property.py, but had enough minor differences that I thought it would be better to WET rather than trying to shoehorn in those differences.

Reply:
"a little duplication is better than the wrong abstraction" 👍